### PR TITLE
Fix redirected_by_self check for dual stack sockets

### PR DESCRIPTION
**Pull request by saxena-anurag**
## Description

In the case when the original connection is using a dual stack socket (hence it invokes a v6 connect_redirect filter), and the new connection initiated by the proxy is using v4 socket, check for redirect state fails, as the filter ID for these 2 filters and hence the redirect_handle is different. Due to that `FwpsQueryConnectionRedirectState` returns REDIRECTED_BY_OTHER which causes the connection to be redirected again.

The fix changes the logic of `redirect_handle` allocation as follows: `redirect_handle` is now allocated per `filter_context`, and not per `filter`. So one `redirect_handle` is used per eBPF program attached to the SOCK_ADDR_CONNECT hook.

This PR also has 2 bug fixes:
1. Call `FwpsRedirectHandleDestroy` to close the redirect handles. UM mock is enhanced to catch any such leak.
2. Move initialization of `sock_addr_ctx` up in connect_redirect_classify, as sock_addr_ctx->compartmentId is used in multiple places in that function.

## Testing

1. Existing CICD
2. Manual testing of the proxy / redirection scenario with dual stack sockets.

## Documentation

NA

### CODE REVIEW

File: netebpfext/net_ebpf_ext_sock_addr.c [hunk 1]
```diff
@@ -1507,6 +1561,7 @@ net_ebpf_extension_sock_addr_redirect_connection_classify(
             NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
             "No \"write\" right; exiting.");
         verdict = BPF_SOCK_ADDR_VERDICT_PROCEED;
+        NET_EBPF_EXT_LOG_ENTRY();
```
Comment (shankarseal) [inline netebpfext/net_ebpf_ext_sock_addr.c:1564]: typo.

**Review by shankarseal — APPROVED** [review]

File: netebpfext/net_ebpf_ext_sock_addr.c [hunk 2]
```diff
@@ -1521,6 +1445,11 @@ net_ebpf_extension_sock_addr_redirect_connection_classify(
         goto Exit;
     }
 
+    // Populate the sock_addr context with WFP classify input fields.
```
Comment (shankarseal) [inline netebpfext/net_ebpf_ext_sock_addr.c:1448]: what was the need to move this up?

Comment (saxena-anurag) [reply netebpfext/net_ebpf_ext_sock_addr.c:1448]: This was a bug. This function is accessing sock_addr struct at many places even before it was initialized (but it was harmless till now as it is only accessing the compartment id from the struct). If we attach multiple programs to different compartments, we start hitting an assert in the code

**Comment by dthaler** [comment]
@saxena-anurag Is there no way to add a test to the automated tests?

**Comment by saxena-anurag** [comment]
> @saxena-anurag Is there no way to add a test to the automated tests?

Filed #2574 for adding automated tests for the proxy scenario.

File: netebpfext/net_ebpf_ext_sock_addr.c (hunk 2 above)
Comment (shankarseal) [reply netebpfext/net_ebpf_ext_sock_addr.c:1448]: Please file test gap issue.

**Review by shankarseal — APPROVED** [review]

File: netebpfext/net_ebpf_ext_sock_addr.c (hunk 2 above)
Comment (saxena-anurag) [reply netebpfext/net_ebpf_ext_sock_addr.c:1448]: Already filed #2574

**Review by dthaler — APPROVED** [review]
Approving because an issue was filed for automated tests.
The guidance though is automated tests should be in the same PR so this shouldn't be taken as a precedent, but rather an exception.

**Pull request merged**